### PR TITLE
Missing private link DNS zones

### PR DESCRIPTION
**Pull request by bentaylorwork**
## Overview/Summary

Adds Private Endpoint DNS zones for Purview and Synapse. The DNS zones are present in the ARM version of the ESLZ with a reference to them here:

https://github.com/Azure/Enterprise-Scale/blob/3b35eb22471e23436e5a0bb4cddc224b2edec686/eslzArm/eslzArm.json#L862

## This PR fixes/adds/changes/removes

1. Adds private DNS zones for the following data related private endpoints:
* privatelink.azuresynapse.net
* privatelink.dev.azuresynapse.net
* privatelink.purview.azure.com
* privatelink.purviewstudio.azure.com

## Testing Evidence

Terraform plans and apply work locally

## As part of this Pull Request I have

- [X ] Checked for duplicate [Pull Requests](https://github.com/Azure/terraform-azurerm-caf-enterprise-scale/pulls)
- [ ] Associated it with relevant [issues](https://github.com/Azure/terraform-azurerm-caf-enterprise-scale/issues), for tracking and closure.
- [X] Ensured my code/branch is up-to-date with the latest changes in the `main` [branch](https://github.com/Azure/terraform-azurerm-caf-enterprise-scale/tree/main)
- [X] Performed testing and provided evidence.
- [X] Updated relevant and associated documentation.
- [ ] Updated the ["What's New?"](https://github.com/Azure/Enterprise-Scale/wiki/Whats-new) wiki page (located in the [Enterprise-Scale repo](https://github.com/Azure/Enterprise-Scale) in the directory: `/docs/wiki/whats-new.md`)

### CODE REVIEW

**Comment by krowlandson** [comment]
Thank you for raising this @bentaylorwork 

I will test and review the code changes, but just wanted to check whether you would be willing to add any other missing services as documented in the [Azure services DNS zone configuration](https://learn.microsoft.com/en-us/azure/private-link/private-endpoint-dns#azure-services-dns-zone-configuration) document?

This will be a breaking change due to the associated input variable schema update so before we release this we will need to get them all updated.

There are also some interesting new additions to these, specifically the new static web apps URI which requires a `partitionId` value:

image

Although we could support this, it will require some additional configuration settings adding to take a list of required `parititionId` values.

As we have limited policy coverage in this space, an alternative approach may be to exclude this one and ask users to provide the full URI using the [config.private_dns_zones]() input variable.

We might also be better including this as a broader set of changes as we move to Terraform `v1.3.0` so we can include these additional inputs `optional()` to reduce impact on existing customers.

Thoughts please @sitarant @jtracey93 @matt-FFFFFF ?

File: modules/connectivity/locals.tf [hunk 1]
```diff
@@ -1288,6 +1288,8 @@ locals {
     azure_sql_database_sqlserver         = ["privatelink.database.windows.net"]
     azure_synapse_analytics_sqlserver    = ["privatelink.database.windows.net"]
     azure_synapse_analytics_sql          = ["privatelink.sql.azuresynapse.net"]
+    azure_synapse                        = ["privatelink.azuresynapse.net"]
```
Comment (krowlandson) [inline modules/connectivity/locals.tf:1291]: Would this be better referred to as `azure_synapse_studio`?

@mboswell and @marvinbuss please can you input on this one, but also a broader view on the current state of our data product coverage? i.e. are the `azure_synapse_analytics_sqlserver` and `azure_synapse_analytics_sql` entries still relevant as the former no longer seems to be mentioned in the docs? 😄

**Review by krowlandson — CHANGES_REQUESTED** [review]
This looks great, and thank you for your contribution @bentaylorwork 🚀 

Please take a look at the feedback and let me know if you would be happy to add to this or would need someone else to add the other services as mentioned in my [PR comment](https://github.com/Azure/terraform-azurerm-caf-enterprise-scale/pull/481#issuecomment-1271493137)

**Comment by krowlandson** [comment]
I've done a quick review of the DNS zone diff and see the following which we should include in a release:

| Private link resource type / Subresource | Status |
| --- | --- |
| Azure Automation / (Microsoft.Automation/automationAccounts) / Webhook, DSCAndHybridWorker | no changes needed |
| Azure SQL Database (Microsoft.Sql/servers) / sqlServer | no changes needed |
| Azure SQL Managed Instance (Microsoft.Sql/managedInstances) | needs testing to verify works with `privatelink.{dnsPrefix}.database.windows.net` format |
| Azure Synapse Analytics (Microsoft.Synapse/workspaces) / Sql | no changes needed |
| Azure Synapse Analytics (Microsoft.Synapse/workspaces) / SqlOnDemand | missing, but same zone name as `Sql` subresource |
| Azure Synapse Analytics (Microsoft.Synapse/workspaces) / Dev | included in this PR |
| Azure Synapse Studio (Microsoft.Synapse/privateLinkHubs) / Web | included in this PR |
| Storage account (Microsoft.Storage/storageAccounts) / Blob (blob, blob_secondary) | no changes needed |
| Storage account (Microsoft.Storage/storageAccounts) / Table (table, table_secondary) | no changes needed |
| Storage account (Microsoft.Storage/storageAccounts) / Queue (queue, queue_secondary) | no changes needed |
| Storage account (Microsoft.Storage/storageAccounts) / File (file, file_secondary) | no changes needed |
| Storage account (Microsoft.Storage/storageAccounts) / Web (web, web_secondary) | no changes needed |
| Azure Data Lake File System Gen2 (Microsoft.Storage/storageAccounts) / Data Lake File System Gen2 (dfs, dfs_secondary) | no changes needed |
| Azure Cosmos DB (Microsoft.AzureCosmosDB/databaseAccounts) / Sql | no changes needed |
| Azure Cosmos DB (Microsoft.AzureCosmosDB/databaseAccounts) / MongoDB | no changes needed |
| Azure Cosmos DB (Microsoft.AzureCosmosDB/databaseAccounts) / Cassandra | no changes needed |
| Azure Cosmos DB (Microsoft.AzureCosmosDB/databaseAccounts) / Gremlin | no changes needed |
| Azure Cosmos DB (Microsoft.AzureCosmosDB/databaseAccounts) / Table | no changes needed |
| Azure Batch (Microsoft.Batch/batchAccounts) / batchAccount | **missing** |
| Azure Batch (Microsoft.Batch/batchAccounts) / nodeManagement | **missing** |
| Azure Database for PostgreSQL - Single server (Microsoft.DBforPostgreSQL/servers) / postgresqlServer | no changes needed |
| Azure Database for MySQL (Microsoft.DBforMySQL/servers) / mysqlServer | no changes needed |
| Azure Database for MariaDB (Microsoft.DBforMariaDB/servers) / mariadbServer | no changes needed |
| Azure Key Vault (Microsoft.KeyVault/vaults) / vault | no changes needed |
| Azure Key Vault (Microsoft.KeyVault/managedHSMs) / Managed HSMs | **missing** |
| Azure Kubernetes Service - Kubernetes API (Microsoft.ContainerService/managedClusters) / management | need to validate region format is correct and check requirements for `{subzone}.privatelink.{region}.azmk8s.io` zone |
| Azure Search (Microsoft.Search/searchServices) / searchService | no changes needed |
| Azure Container Registry (Microsoft.ContainerRegistry/registries) / registry | need to test whether regional zones work as expected for `{region}.privatelink.azurecr.io` |
| Azure App Configuration (Microsoft.AppConfiguration/configurationStores) / configurationStores | no changes needed |
| Azure Backup (Microsoft.RecoveryServices/vaults) / AzureBackup | no changes needed |
| Azure Site Recovery (Microsoft.RecoveryServices/vaults) / AzureSiteRecovery | need to check as zone is now documented as being regional, i.e. `privatelink.{region}.siterecovery.windowsazure.com` |
| Azure Event Hubs (Microsoft.EventHub/namespaces) / namespace | no changes needed |
| Azure Service Bus (Microsoft.ServiceBus/namespaces) / namespace | no changes needed |
| Azure IoT Hub (Microsoft.Devices/IotHubs) / iotHub | no changes needed |
| Azure Relay (Microsoft.Relay/namespaces) / namespace | no changes needed |
| Azure Event Grid (Microsoft.EventGrid/topics) / topic | no changes needed |
| Azure Event Grid (Microsoft.EventGrid/domains) / domain | no changes needed |
| Azure Web Apps (Microsoft.Web/sites) / sites | no changes needed |
| Azure Machine Learning (Microsoft.MachineLearningServices/workspaces) / amlworkspace | no changes needed |
| SignalR (Microsoft.SignalRService/SignalR) / signalR | no changes needed |
| Azure Monitor (Microsoft.Insights/privateLinkScopes) / azuremonitor | missing `privatelink.blob.core.windows.net` zone |
| Cognitive Services (Microsoft.CognitiveServices/accounts) / account | no changes needed |
| Azure File Sync (Microsoft.StorageSync/storageSyncServices) / afs | need to check as zone is now documented as being regional, i.e. `privatelink.{region}.afs.azure.net` |
| Azure Data Factory (Microsoft.DataFactory/factories) / dataFactory | no changes needed |
| Azure Data Factory (Microsoft.DataFactory/factories) / portal | no changes needed |
| Azure Cache for Redis (Microsoft.Cache/Redis) / redisCache | no changes needed |
| Azure Cache for Redis Enterprise (Microsoft.Cache/RedisEnterprise) / redisEnterprise | **missing** |
| Microsoft Purview (Microsoft.Purview) / account | included in this PR |
| Microsoft Purview (Microsoft.Purview) / portal | included in this PR |
| Azure Digital Twins (Microsoft.DigitalTwins) / digitalTwinsInstances | **missing** |
| Azure HDInsight (Microsoft.HDInsight) | **missing** |
| Azure Arc (Microsoft.HybridCompute) / hybridcompute | **missing** |
| Azure Media Services (Microsoft.Media) / keydelivery, liveevent, streamingendpoint | **missing** |
| Azure Data Explorer (Microsoft.Kusto) | **missing** |
| Azure Static Web Apps (Microsoft.Web/staticSites) / staticSites | **missing** |
| Azure Migrate (Microsoft.Migrate) / migrate projects, assessment project and discovery site | **missing** |
| Azure Managed HSM (Microsoft.Keyvault/managedHSMs) / managedhsm | **missing** |
| Azure API Management (Microsoft.ApiManagement/service) / gateway | **missing** |
| Microsoft PowerBI (Microsoft.PowerBI/privateLinkServicesForPowerBI) | **missing** |
| Azure Bot Service (Microsoft.BotService/botServices) / Bot | **missing** |
| Azure Bot Service (Microsoft.BotService/botServices) / Token | **missing** |

File: docs/wiki/[User-Guide]-Module-Variables.md [hunk 2]
```diff
@@ -42,7 +42,7 @@ Default: `{}`
...
...ngodb = bool azure_cosmos_db_cassandra = bool azure_cosmos_db_gremlin = bool azure_cosmos_db_table = bool azure_database_for_postgresql_server = bool azure_database_for_mysql_server = bool azure_database_for_mariadb_server = bool azure_key_vault = bool azure_kubernetes_service_management = bool azure_search_service = bool azure_container_registry = bool azure_app_configuration_stores = bool azure_backup = bool azure_site_recovery = bool azure_event_hubs_namespace = bool azure_service_bus_namespace = bool azure_iot_hub = bool azure_relay_namespace = bool azure_event_grid_topic = bool azure_event_grid_domain = bool azure_web_apps_sites = bool azure_machine_learning_workspace = bool signalr = bool azure_monitor = bool cognitive_services_account = bool azure_file_sync = bool azure_data_factory = bool azure_data_factory_portal = bool azure_cache_for_redis = bool azure_purview = bool azure_purview_studio = bool }) private_link_locations = list(string) public_dns_zones = list(string) private_dns_zones = list(string) enable_private_dns_zone_virtual_network_link_on_hubs = bool enable_private_dns_zone_virtual_network_link_on_spokes = bool }) }) }) location = any tags = any advanced = any })`
```
[the remark this answers is not in the document]
Comment (bentaylorwork) [reply docs/wiki/[User-Guide]-Module-Variables.md]: Should be resolved in latest commit.

File: docs/wiki/[Variables]-configure_connectivity_resources.md [hunk 3]
```diff
@@ -119,47 +119,51 @@ If specified, will customize the "connectivity" landing zone settings and resour
...
-          azure_web_apps_sites                 = true
-          azure_machine_learning_workspace     = true
-          signalr                              = true
-          azure_monitor                        = true
-          cognitive_services_account           = true
-          azure_file_sync                      = true
-          azure_data_factory                   = true
-          azure_data_factory_portal            = true
-          azure_cache_for_redis                = true
+            azure_automation_webhook             = true
```
[the remark this answers is not in the document]
Comment (bentaylorwork) [reply docs/wiki/[Variables]-configure_connectivity_resources.md]: Should be resolved in latest commit.

**Comment by bentaylorwork** [comment]
I can take a look at adding some of the missing Private DNS Zones early next week.

Do you want me to add them to this PR or open a new one?

**Comment by krowlandson** [comment]
> I can take a look at adding some of the missing Private DNS Zones early next week.
> 
> Do you want me to add them to this PR or open a new one?

That's great, thank you. I think it would be good to include in a single PR so we know what they are all covered off when merging to main and cutting our next release.

Something else that will need updating is the test framework settings here:

https://github.com/Azure/terraform-azurerm-caf-enterprise-scale/blob/5a54e8629c6a5dde00dc54d5c63b664b37e2881f/tests/modules/settings/settings.connectivity.tf#L201-L243

🪨

**Comment by bentaylorwork** [comment]
I have dealt with all the private link dns zones apart from the ones in bold below. I will try and get to them in the next few days.

Private link resource type / Subresource | Status
-- | --
Azure Automation / (Microsoft.Automation/automationAccounts) / Webhook, DSCAndHybridWorker | no changes needed
Azure SQL Database (Microsoft.Sql/servers) / sqlServer | no changes needed
Azure SQL Managed Instance (Microsoft.Sql/managedInstances) | **needs testing to verify works with privatelink.{dnsPrefix}.database.windows.net format**
Azure Synapse Analytics (Microsoft.Synapse/workspaces) / Sql | no changes needed
Azure Synapse Analytics (Microsoft.Synapse/workspaces) / SqlOnDemand | **missing, but same zone name as Sql subresource**
Azure Synapse Analytics (Microsoft.Synapse/workspaces) / Dev | included in this PR
Azure Synapse Studio (Microsoft.Synapse/privateLinkHubs) / Web | included in this PR
Storage account (Microsoft.Storage/storageAccounts) / Blob (blob, blob_secondary) | no changes needed
Storage account (Microsoft.Storage/storageAccounts) / Table (table, table_secondary) | no changes needed
Storage account (Microsoft.Storage/storageAccounts) / Queue (queue, queue_secondary) | no changes needed
Storage account (Microsoft.Storage/storageAccounts) / File (file, file_secondary) | no changes needed
Storage account (Microsoft.Storage/storageAccounts) / Web (web, web_secondary) | no changes needed
Azure Data Lake File System Gen2 (Microsoft.Storage/storageAccounts) / Data Lake File System Gen2 (dfs, dfs_secondary) | no changes needed
Azure Cosmos DB (Microsoft.AzureCosmosDB/databaseAccounts) / Sql | no changes needed
Azure Cosmos DB (Microsoft.AzureCosmosDB/databaseAccounts) / MongoDB | no changes needed
Azure Cosmos DB (Microsoft.AzureCosmosDB/databaseAccounts) / Cassandra | no changes needed
Azure Cosmos DB (Microsoft.AzureCosmosDB/databaseAccounts) / Gremlin | no changes needed
Azure Cosmos DB (Microsoft.AzureCosmosDB/databaseAccounts) / Table | no changes needed
Azure Batch (Microsoft.Batch/batchAccounts) / batchAccount | included in this PR
Azure Batch (Microsoft.Batch/batchAccounts) / nodeManagement | included in this PR
Azure Database for PostgreSQL - Single server (Microsoft.DBforPostgreSQL/servers) / postgresqlServer | no changes needed
Azure Database for MySQL (Microsoft.DBforMySQL/servers) / mysqlServer | no changes needed
Azure Database for MariaDB (Microsoft.DBforMariaDB/servers) / mariadbServer | no changes needed
Azure Key Vault (Microsoft.KeyVault/vaults) / vault | no changes needed
Azure Key Vault (Microsoft.KeyVault/managedHSMs) / Managed HSMs | included in this PR
Azure Kubernetes Service - Kubernetes API (Microsoft.ContainerService/managedClusters) / management | need to validate region format is correct and check requirements for {subzone}.privatelink.{region}.azmk8s.io zone
Azure Search (Microsoft.Search/searchServices) / searchService | no changes needed
Azure Container Registry (Microsoft.ContainerRegistry/registries) / registry | **need to test whether regional zones work as expected for {region}.privatelink.azurecr.io**
Azure App Configuration (Microsoft.AppConfiguration/configurationStores) / configurationStores | no changes needed
Azure Backup (Microsoft.RecoveryServices/vaults) / AzureBackup | no changes needed
Azure Site Recovery (Microsoft.RecoveryServices/vaults) / AzureSiteRecovery | **need to check as zone is now documented as being regional, i.e. privatelink.{region}.siterecovery.windowsazure.com**
Azure Event Hubs (Microsoft.EventHub/namespaces) / namespace | no changes needed
Azure Service Bus (Microsoft.ServiceBus/namespaces) / namespace | no changes needed
Azure IoT Hub (Microsoft.Devices/IotHubs) / iotHub | no changes needed
Azure Relay (Microsoft.Relay/namespaces) / namespace | no changes needed
Azure Event Grid (Microsoft.EventGrid/topics) / topic | no changes needed
Azure Event Grid (Microsoft.EventGrid/domains) / domain | no changes needed
Azure Web Apps (Microsoft.Web/sites) / sites | no changes needed
Azure Machine Learning (Microsoft.MachineLearningServices/workspaces) / amlworkspace | no changes needed
SignalR (Microsoft.SignalRService/SignalR) / signalR | no changes needed
Azure Monitor (Microsoft.Insights/privateLinkScopes) / azuremonitor | **Looks like this is all ready included under the var storage_account_blob **
Cognitive Services (Microsoft.CognitiveServices/accounts) / account | no changes needed
Azure File Sync (Microsoft.StorageSync/storageSyncServices) / afs | **need to check as zone is now documented as being regional, i.e. privatelink.{region}.afs.azure.net**
Azure Data Factory (Microsoft.DataFactory/factories) / dataFactory | no changes needed
Azure Data Factory (Microsoft.DataFactory/factories) / portal | no changes needed
Azure Cache for Redis (Microsoft.Cache/Redis) / redisCache | no changes needed
Azure Cache for Redis Enterprise (Microsoft.Cache/RedisEnterprise) / redisEnterprise | included in this PR
Microsoft Purview (Microsoft.Purview) / account | included in this PR
Microsoft Purview (Microsoft.Purview) / portal | included in this PR
Azure Digital Twins (Microsoft.DigitalTwins) / digitalTwinsInstances | included in this PR
Azure HDInsight (Microsoft.HDInsight) | included in this PR
Azure Arc (Microsoft.HybridCompute) / hybridcompute | included in this PR
Azure Media Services (Microsoft.Media) / keydelivery, liveevent, streamingendpoint | included in this PR
Azure Data Explorer (Microsoft.Kusto) | included in this PR
Azure Static Web Apps (Microsoft.Web/staticSites) / staticSites | **missing - https://github.com/MicrosoftDocs/azure-docs/issues/93952 - might be best to get the user to provide a list of partition ids then when a full list is available it can be added in and merged with the user provided list - thoughts?**
Azure Migrate (Microsoft.Migrate) / migrate projects, assessment project and discovery site | included in this PR
Azure Managed HSM (Microsoft.Keyvault/managedHSMs) / managedhsm | included in this PR
Azure API Management (Microsoft.ApiManagement/service) / gateway | included in this PR
Microsoft PowerBI (Microsoft.PowerBI/privateLinkServicesForPowerBI) | included in this PR
Azure Bot Service (Microsoft.BotService/botServices) / Bot | included in this PR
Azure Bot Service (Microsoft.BotService/botServices) / Token | included in this PR

**Comment by krowlandson** [comment]
> I have dealt with all the private link dns zones apart from the ones in bold below. I will try and get to them in the next few days.

This is awesome stuff... thank you @bentaylorwork !!

I will try to make some time tomorrow to provide feedback on current progress as a couple of minor things I would like to change for consistency, but also to clarify some of the thinking around where services share an underlying namespace 👍🏻

File: modules/connectivity/locals.tf (hunk 1 above)
Comment (marvinbuss) [reply modules/connectivity/locals.tf:1291]: Hi @krowlandson,
Sorry, I just saw that you tagged me in this comment. `privatelink.database.windows.net` is not relevant for Synapse. This is only relevant for Azure SQL Servers.

Synapse only requires:
1. For SQL endpoints: `privatelink.sql.azuresynapse.net`
2. For dev endpoints: `privatelink.dev.azuresynapse.net`
3. For Data Explorer endpoints: `privatelink.kusto.azuresynapse.net`
4. For Synapse Studio: `privatelink.azuresynapse.net`

Feel free to rename one of the variables to `azure_synapse_studio` accordingly.

Comment (marvinbuss) [reply modules/connectivity/locals.tf:1291]: Hi @krowlandson,
Sorry, I just saw that you tagged me in this comment. `privatelink.database.windows.net` is not relevant for Synapse. This is only relevant for Azure SQL Servers.

Synapse only requires:
1. For SQL endpoints: `privatelink.sql.azuresynapse.net`
2. For dev endpoints: `privatelink.dev.azuresynapse.net`
3. For Data Explorer endpoints: `privatelink.kusto.azuresynapse.net`
4. For Synapse Studio: `privatelink.azuresynapse.net`

Feel free to rename one of the variables to `azure_synapse_studio` accordingly.

Comment (krowlandson) [reply modules/connectivity/locals.tf:1291]: Cool, thank you for the confirmation @marvinbuss 👍🏻 

I thought it was odd that the docs were updated, as the module reflects what the docs said at the time of authoring this capability, so we definitely need to fix this then 😄

**Comment by krowlandson** [comment]
@bentaylorwork ... I've made some updates so we can keep this progressing, however I need to validate why this is happening on our `Update` task:

image

I believe it's due to your PR coming from the `main` branch in your fork.

Working locally, I observed the same permissions error when trying to push to just `HEAD`. To resolve this I had to push to `HEAD:main` on your fork so I may be able to fix this with an update to our push logic. Please bear with me while I look into this!

**Review by krowlandson — APPROVED** [review]
Although we have a few further DNS related updates we want to implement, I think this is a great start and gets us past a significant milestone in this work.

Thank you for your contribution @bentaylorwork 💯 

LGTM 🚀

**Pull request merged**